### PR TITLE
[ONNX] Redesign inplace conversion

### PR DESCRIPTION
**Pull request by BowenBao**
* Create `InplaceConverter` and `ValueTracker` to keep track of aliases of values throughout the graph. For a given value, a new alias is created every time when there is an inplace operation, SetAttr, or through nested blocks owned by If/Loop nodes.
* Fix bug where controlflow node output types are not set, when the complete node is unable to run ONNX shape inference due to containing non-onnx node. 
* Add symbolic for `__not__` ~~and `prim_min`~~(update: moved to a separate PR), and update `index_put` opset9 to support case of assignment without providing indices. 
* Bump ORT version in CI test.

### CODE REVIEW

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 1]
```diff
@@ -567,256 +370,390 @@ Value* findArgumentAsInputParam(
...
+
+    std::string fullName("");
+    for (auto& name : moduleNames) {
+      fullName += name + '.';
+    }
+    fullName += name;
+
+    attr_node_fullname_map.insert({n, fullName});
+
+    if (attr_name_value_map.find(fullName) == attr_name_value_map.end() &&
```
Comment (shubhambhokare1) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:601]: nit: Anyway we can combine this if and L#630 if, in order to avoid trying to find fullname twice

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 2]
```diff
@@ -567,256 +370,390 @@ Value* findArgumentAsInputParam(
...
         }
+      } else if (nkind == aten::insert || nkind == aten::append) {
+        std::tie(orig_data, new_out) = PrepareListAppendAndInsertForONNX(n);
+      } else if (mr_->inplaceOpVariant(n)) {
+        std::tie(orig_data, new_out) = PrepareInplaceOpsInBlocksForONNX(n);
+      } else if (nkind == aten::pop) {
+        std::tie(orig_data, new_out) = PrepareListPopForONNX(n);
+      } else if (nkind == aten::Delete) {
+        std::tie(orig_data, new_out) = PrepareListDeleteForONNX(n);
+      } else if (nkind == aten::_set_item) {
```
Comment (shubhambhokare1) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:727]: I see all the setAttr and getAttr have been converted to inplace ops, I was wondering what exactly is aten::_set_item

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 3]
```diff
@@ -19,17 +19,95 @@ namespace {
...
+    void recordSetValue(Value* old_v, Value* new_v);
+    Value* findAliasForValueAtNode(Value* v, const Node* n) const;
+
+    std::string toString() const;
+
+   private:
+
+    std::shared_ptr<Graph> graph_;
+
+    // Map from aliases to root value.
```
Comment (shubhambhokare1) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:67]: nit: What exactly is an alias? Maybe you could add an explanation here?

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 4]
```diff
@@ -19,17 +19,95 @@ namespace {
...
+      bool operator()(const Value* a, const Value* b) const {
+        auto* n_a = a->node();
+        auto* n_b = b->node();
+        if (n_a == n_b) {
+          return false;
+        }
+        auto a_b = n_a->isBefore(n_b);
+        auto b_a = n_b->isBefore(n_a);
+        if (a_b == b_a) {
+          return a->unique() < b->unique();
```
Comment (shubhambhokare1) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:87]: Why is the unique id used as a tiebreaker?

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 5]
```diff
@@ -567,256 +370,390 @@ Value* findArgumentAsInputParam(
...
+        isAncestor(alias_n->owningBlock(), n->owningBlock())) {
+      found_alias = alias;
+    }
+  }
+
+  TORCH_INTERNAL_ASSERT(nullptr != found_alias, "More details: \n",
+    n->sourceRange().str(),
+    "Input ", v->debugName(), " of node ", *n,
+    " was modified by in-place operation, but we cannot find its updated value. ",
+    "Please report a bug to PyTorch, and/or try to avoid using in-place operators on this value.");
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:563]: what situations can cause this? Why does the user need to report a bug?

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 6]
```diff
@@ -567,256 +370,390 @@ Value* findArgumentAsInputParam(
...
+      } else {
+        GRAPH_DEBUG(
+            attr.type()->cast<ClassType>() ? "" : "attribute: ",
+            name,
+            " is not materializable.");
+      }
+    }
+
+    // Create dummy initial value.
+    if (attr_name_value_map.find(fullName) == attr_name_value_map.end()) {
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:630]: When is this part needed?

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 7]
```diff
@@ -567,256 +370,390 @@ Value* findArgumentAsInputParam(
...
+      auto* noneNode = graph_->create(prim::Constant);
+      noneNode->output()->setType(NoneType::get());
+      noneNode->insertBefore(graph_->nodes().front());
+      attr_name_value_map.insert({fullName, noneNode->output()});
+    }
+  }
+}
+
+// Replace prim::GetAttr and prim::SetAttr with ATen inplace operators.
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:639]: Can you add a graph example here in comments?

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 8]
```diff
@@ -567,256 +370,390 @@ Value* findArgumentAsInputParam(
...
+      auto new_block_in = owning_block->addInput();
+      sorted_alias.insert(new_block_in);
+      alias_to_value_[new_block_in] = root_v;
+      owning_blocknode->addInput(root_v);
+    } else if (owning_block_nkind == prim::If) {
+      for (auto* if_sub_block : owning_blocknode->blocks()) {
+        if (owning_block == if_sub_block) {
+          if_sub_block->registerOutput(new_v);
+        } else {
+          if_sub_block->registerOutput(root_v);
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:473]: Can we register an output from outside of the block? Do we need a clone?

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 3 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:67]: it is explained in the line below, `A single value can have multiple aliases throughout the graph, created by inplace operators, ...`

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 4 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:87]: unique id is guaranteed to be different for different `Value` obj. Ties must be broken, otherwise the two aliases will be considered equal to each other, and then only one of them gets stored into the map.

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 8 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:473]: Yes we can, clone(Identity) is added in `fixup_onnx_controlflow` pass

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 5 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:563]: unknown cases :) This is not expected to happen, but if it does, the hope is that user can get some meaningful message explaining what might have gone wrong.

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 6 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:630]: For handling the case where code enters the else branch on L#622. I'm not familiar with that case actually, is it possible to reach there?

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 2 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:727]: created by `tensor_list[i] = some_tensor`

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 1 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:601]: L#630 is to check before leaving this function, if `fullname` exist in `attr_name_value_map`, and insert dummy initial value if that's not the case. So the outcome of the two finds is different. But yes there is room to improve the overall logic here.

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp [hunk 9]
```diff
@@ -236,6 +236,11 @@ std::vector<Value*> FixupONNXLoopNode(Node* node, int opset_version) {
   // NOTE: the output order is deliberately changed to match expected order
   //       since onnx loop requires scan outputs to be the last outputs.
   auto new_outputs = ConvertSequenceDependencies(node, opset_version);
+
+  // TODO: fix node output metadata based on block output metadata.
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp:240]: good catch, i was updating another branch on this part of the code :(

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 10]
```diff
@@ -4353,6 +4353,36 @@ def forward(self, x, y):
         y = torch.randn(4, 5)
         self.run_test(model, (x, y))
 
+    @skipIfUnsupportedMinOpsetVersion(11)
+    def test_list_set(self):
+        class ListModel(torch.nn.Module):
+            def forward(self, x, y):
+                res = []
+                for i in range(x.size(0)):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:4361]: sure, added

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 7 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:641]: added

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:4379]: Question: In case of append in If subblock, doesn't the other block need to clone the list? Does it need to use Identity on sequence?

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp]: Sorry I didn't get it, is it intended? Same for line 383

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:4379]: identity will be inserted at fixup_onnx_controlflow, yes it will use identity on sequence.

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 6 above)
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:636]: There are cases that attrModule.hasattr(name) could be false, but I'm trying to remember. 
Some attributes are actually not registered in some modules. Like maybe num_batches_tracked for BatchNorm? Can we test this?

File: torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/fixup_onnx_controlflow.cpp]: no, only the word `TODO` was not intended, the rest are.

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp (hunk 6 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:636]: Was num_batches_tracked covered before? I couldn't repro this with the current batch_norm training case.

**Review by neginraoof — APPROVED** [review]
LGTM. Thanks!
